tariff/ostrom.go

```diff
@@ -0,0 +1,276 @@
+package tariff
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"slices"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/tariff/ostrom"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/oauth"
+	"github.com/evcc-io/evcc/util/request"
+	"github.com/evcc-io/evcc/util/transport"
+	"github.com/jinzhu/now"
+	"golang.org/x/oauth2"
+)
+
+type Ostrom struct {
+	*embed
+	*request.Helper
+	log          *util.Logger
+	zip          string
+	contractType string
+	cityId       int // Required for the Fair tariff types
+	basic        string
+	data         *util.Monitor[api.Rates]
+}
+
+var _ api.Tariff = (*Ostrom)(nil)
+
+func init() {
+	registry.Add("ostrom", NewOstromFromConfig)
+}
+
+// Search for a contract in list of contracts
+func ensureContractEx(cid int64, contracts []ostrom.Contract) (ostrom.Contract, error) {
+	var zero ostrom.Contract
+
+	if cid != -1 {
+		// cid defined
+		for _, contract := range contracts {
+			if cid == contract.Id {
+				return contract, nil
+			}
+		}
+	} else if len(contracts) == 1 {
+		// cid empty and exactly one object
+		return contracts[0], nil
+	}
+
+	return zero, fmt.Errorf("cannot find contract")
+}
+
+func NewOstromFromConfig(other map[string]interface{}) (api.Tariff, error) {
+	var cc struct {
+		ClientId     string
+		ClientSecret string
+		Contract     int64
+	}
+
+	cc.Contract = -1
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	if cc.ClientId == "" || cc.ClientSecret == "" {
+		return nil, api.ErrMissingCredentials
+	}
+
+	basic := transport.BasicAuthHeader(cc.ClientId, cc.ClientSecret)
+	log := util.NewLogger("ostrom").Redact(basic)
+
+	t := &Ostrom{
+		log:          log,
+		basic:        basic,
+		contractType: ostrom.PRODUCT_DYNAMIC,
```

Comment:
why is this needed?

Reply:
This is needed to implement the api.TarrifType interface.
Ostrom offers two different types of contracts, one with dynamic and one with (more or less) static prices. This stores the information which one it is.

Reply:
Doesnt that already happen in https://github.com/evcc-io/evcc/pull/16354/files#diff-91620eb465981824c351f0d81e5ab643b5b3751232cce50d8ee91d541db1091eR101?

Reply:
Sorry, but I don't understand which part of the code you refer to.
The Function `Type()` uses the value stored in this member to decide on the type of the tariff. The initialization here is not strictly necessary, since the value will be set after the contracts have been read.